src/coreclr/jit/forwardsub.cpp

```diff
@@ -0,0 +1,737 @@
+    // See if the next statement contains the one and only use.
+    //
+    Statement* const nextStmt = stmt->GetNextStmt();
+
+    // We often see stale flags, eg call flags after inlining.
+    // Try and clean these up.
+    //
+    gtUpdateStmtSideEffects(nextStmt);
+    gtUpdateStmtSideEffects(stmt);
+
+    // Scan for the (single) use.
+    //
+    ForwardSubVisitor fsv(this, lclNum);
+    fsv.WalkTree(nextStmt->GetRootNodePointer(), nullptr);
+
+    // LclMorph (via RCS_Early) said there was just one use.
+    // It had better have gotten this right.
+    //
+    assert(fsv.GetUseCount() <= 1);
+
+    if ((fsv.GetUseCount() == 0) || (fsv.GetNode() == nullptr))
+    {
+        JITDUMP(" no next stmt use\n");
+        return false;
+    }
+
+    JITDUMP(" [%06u] is only use of [%06u] (V%02u) ", dspTreeID(fsv.GetNode()), dspTreeID(lhsNode), lclNum);
+
+    // If next statement already has a large tree, hold off
+    // on making it even larger.
+    //
+    // We use total node count. Consider instead using the depth of the use and the
+    // height of the fwdSubNode.
+    //
+    unsigned const nextTreeLimit = 200;
+    if ((fsv.GetComplexity() > nextTreeLimit) && gtComplexityExceeds(&fwdSubNode, 1))
+    {
+        JITDUMP(" next stmt tree is too large (%u)\n", fsv.GetComplexity());
+        return false;
+    }
+
+    // Next statement seems suitable.
+    // See if we can forward sub without changing semantics.
+    //
+    GenTree* const nextRootNode = nextStmt->GetRootNode();
+
+    // Bail if types disagree.
+    // Might be able to tolerate these by retyping.
+    //
+    if (fsv.GetNode()->TypeGet() != fwdSubNode->TypeGet())
+    {
+        JITDUMP(" mismatched types (substitution)\n");
+        return false;
+    }
+
+    // We can forward sub if
+    //
+    // the value of the fwdSubNode can't change and its evaluation won't cause side effects,
+    //
+    // or,
+    //
+    // if the next tree can't change the value of fwdSubNode or be impacted by fwdSubNode effects
+    //
+    const bool fwdSubNodeInvariant   = ((fwdSubNode->gtFlags & GTF_ALL_EFFECT) == 0);
+    const bool nextTreeIsPureUpToUse = ((fsv.GetFlags() & (GTF_EXCEPT | GTF_GLOB_REF | GTF_CALL)) == 0);
+    if (!fwdSubNodeInvariant && !nextTreeIsPureUpToUse)
+    {
+        // Fwd sub may impact global values and or reorder exceptions...
+        //
+        JITDUMP(" potentially interacting effects\n");
+        return false;
+    }
+
+    // If we're relying on purity of fwdSubNode for legality of forward sub,
+    // do some extra checks for global uses that might not be reflected in the flags.
+    //
+    if (fwdSubNodeInvariant && ((fsv.GetFlags() & (GTF_CALL | GTF_ASG)) != 0))
+    {
+        EffectsVisitor ev(this);
+        ev.WalkTree(&fwdSubNode, nullptr);
+
+        if ((ev.GetFlags() & GTF_GLOB_REF) != 0)
+        {
+            JITDUMP(" potentially interacting effects (AX locals)\n");
+            return false;
+        }
```

Comment:
Sure. And it would be nice to get these flags set correctly (conservatively) right from the importer so no repairing is needed, just (optional) refinement.